[loyaltypollution] iP

loyaltypollution opened the pull request:
# Tom Task Manager

## Features
- Manage tasks efficiently  
- Add, delete, and mark tasks  
- Save and load tasks from a file  

## Usage
1. Clone the repository:  
   ```sh
   git clone https://github.com/your-repo/tom-task-manager.git
   ```
2. Navigate to the project directory:  
   ```sh
   cd tom-task-manager
   ```
3. Compile the project  
4. Run the application  

## To-Do List
- [x] Implement task addition  
- [x] Implement task deletion  
- [ ] Implement task editing  
- [ ] Add more unit tests  

## Emoji  
Keep your tasks organized! :memo:  

> "The key is not to prioritize what's on your schedule, but to schedule your priorities."  
> — *Stephen Covey*  

[Visit our website](https://example.com/)

### Markdown Formatting
- Use inline code for small code snippets:  
  Example: `print("Hello, World!")`  
- **Bold Text**  
- *Italic Text*  
- ~~Strikethrough Text~~

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

loyaltypollution committed:
Duke.java: Rename chatbot name from Duke to Tom

Changes made:
* File name is renamed from Duke.java to Tom.java
to remove all traces of Duke from the source code.

* Ascii art prining Duke has been changed to Tom.

loyaltypollution committed:
Tom.java: Add skeleton that greets user and exits.

loyaltypollution committed:
Tom.java: Add echo functionality and exit on 'bye' command

Implement functionality to echo user input and terminate when 'bye' is entered.

loyaltypollution committed:
Tom.java: Add task storage and listing functionality

* Chatbot will store up to 100 tasks on nonpersistent storage.
* Includes support for acknowledging added tasks.
* Tasks are listed on command "list".

loyaltypollution committed:
Tom.java: Refactor if-elseif to switch case

Refactored the if-elseif statements into a switch-case structure.
This change makes it easier to add additional commands in the future.

loyaltypollution committed:
Task.java, Tom.java: Encapsulate task details in Task class

Task.java implements task details.

Tom.java abstracts the task-strings into a dedicated Task class.

loyaltypollution committed:
Tom.java: correct format of Ascii Art

loyaltypollution committed:
Tom.java: add mark and unmark commands

loyaltypollution committed:
Tom.java: Add template for todo, deadline, event commands

loyaltypollution committed:
README.md: reword Duke to Tom

loyaltypollution committed:
bugfix: enable adding tasks with spaces

loyaltypollution committed:
encapsulation: create Deadline, Event, Todo classes for each task.

loyaltypollution committed:
feature: Perform I/O redirection testing

loyaltypollution committed:
Tom.java: Add template for delete command

loyaltypollution committed:
Tom.java: Refactor Task array to ArrayList

loyaltypollution committed:
Tom.java: Enable Task deletion

loyaltypollution committed:
test: Add "delete" command to testcases

loyaltypollution committed:
Refactor: Organize task-related classes to  Task package

loyaltypollution committed:
Tom.java: Add save command

loyaltypollution committed:
Test: Add "save" command to testcases

loyaltypollution committed:
.gitignore: Ignore data/tom.txt

loyaltypollution committed:
Merge branch 'branch-Level-7'

loyaltypollution committed:
Bug fix: Add toFileFormatString to Task classes

loyaltypollution committed:
Feat: Refactor codebase to OOP for Bye and Todo commands

BREAKING CHANGE: Refactor codebase to support OOP practices

loyaltypollution committed:
Todo command: Provide ui feedback on task addition

loyaltypollution committed:
ListCommand.java: Implement list command

clx3210 commented on src/main/java/Tom.java:
```diff
@@ -0,0 +1,195 @@
+        System.out.println("What can I do for you?");
+
+        Scanner inputBuffer = new Scanner(System.in);
+        boolean continueReading = true;
```
This seems to violate the boolean variable naming convention from the coding standard. How about renaming it to `shouldContinue`?

loyaltypollution replied:
Thank you for the suggestion. Tom.java has been refactored to support more OOP. This boolean variable has been replaced with "isExit" in accordance with the boolean variable naming convention.

clx3210 commented on src/main/java/Tom.java:
```diff
@@ -0,0 +1,195 @@
+            List<String> inputComponents = Arrays.asList(userInput.split(" "));
+            String command = inputComponents.get(0);
+
+            switch (command) {
```
Shall we unindent the case clauses here?

loyaltypollution replied:
The case clauses have been refactored to Parser.java. This has been fixed in recent commit.

clx3210 commented on src/main/java/tom/task/Todo.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    public String toFileFormatString() {
+        return String.format("T | %s | %s", this.getStatusIcon(), this.getDescription());
```
I like the use of `String.format(...)` here, it looks clean.

loyaltypollution replied:
Thank you, will continue to adopt this in future.

clx3210 commented on src/main/java/Tom.java:
```diff
@@ -0,0 +1,195 @@
+                    if (tasks.size() < 100) {
+                        int fromPos = inputComponents.indexOf("/from") + 1;
+                        int toPos = inputComponents.indexOf("/to") + 1;
+                        List<String> fromString = inputComponents.subList(fromPos, toPos);
```
Shall we use plural form naming instead for the `List<String>` objects?

loyaltypollution committed:
Tom.java: Implement Deadline, Event, Mark, Unmark, Save command

clx3210 reviewed (commented):
Looks good! 👍
Just minor comments on naming convention.

loyaltypollution committed:
Tom.java: Implement Delete command

loyaltypollution committed:
Tom.java: Remove unnecessary imports

loyaltypollution committed:
Parser.java: Parse and validate date format for tasks

loyaltypollution committed:
Merge branch 'branch-Level-8'

loyaltypollution committed:
.gitignore: Fix */data/tom.txt -> **/data/tom.txt

loyaltypollution committed:
.gitignore: Ignore IDE config files

loyaltypollution committed:
Parser.java: Fix switch case indentation

loyaltypollution committed:
Merge branch 'add-gradle-support'

loyaltypollution committed:
build.gradle: Update mainClass to Tom

loyaltypollution committed:
test: Add JUnit test for Parser, Todo classes

loyaltypollution committed:
test: Add JUnit test for TaskList classes

loyaltypollution committed:
docs: Add JavaDoc comments to the code

loyaltypollution committed:
chore: Update to comply with coding standard

loyaltypollution committed:
Parser.java: Fix regex for description to allow spaces

loyaltypollution committed:
FindCommand.java: Implement find command

loyaltypollution committed:
test: Add "find" command to testcases

loyaltypollution committed:
Merge branch 'branch-A-JavaDoc'

loyaltypollution committed:
Merge branch 'branch-A-CodingStandard'

loyaltypollution committed:
Merge branch 'branch-Level-9'

loyaltypollution committed:
test: Fix input types for mark, unmark, delete

loyaltypollution committed:
checkstyle: Add checkstyle

loyaltypollution committed:
chore: Fix all code to match checkstyle

loyaltypollution committed:
chore: label usage of variadic arguments

loyaltypollution committed:
Merge branch 'branch-A-Varargs'

loyaltypollution committed:
Merge branch 'branch-A-CheckStyle'

GauhVish commented on src/main/java/Tom.java:
```diff
@@ -0,0 +1,56 @@
+import tom.command.Command;
```
hi missing a package statement up here : )

GauhVish commented on src/main/java/tom/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+ */
+public abstract class Command {
+
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage);
+
+    public abstract boolean isExit();
```
I think you could a javadocs here, even if its an abstract class, though the guide isnt so clear on this

GauhVish commented on src/main/java/tom/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,50 @@
+        return false;
+    }
+
+}
```
Really good job with all the command classes , super nice and neat!

GauhVish commented on src/main/java/tom/parser/Parser.java:
```diff
@@ -0,0 +1,104 @@
+            return new FindCommand(keyword);
+        }
+        default:
+            return new UnknownCommand();
```
Good job on using the right switch case indentation!

GauhVish commented on src/main/java/tom/task/Task.java:
```diff
@@ -0,0 +1,88 @@
+     * @param keyword The keyword to search for in the task description.
+     * @return true if the keyword is found in the description, false otherwise.
+     */
+    public boolean matchKeyword(String keyword) {
```
I think this is good, but maybe something like findKeyword might be a more intuitive name?

GauhVish commented on src/main/java/tom/tasklist/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+     * @return An iterator over the tasks in this list.
+     */
+    @Override
+    public Iterator<Task> iterator() {
```
Hmm , maybe change the name to one with an action, like toIterator ?

GauhVish reviewed (commented):
Hi, i think youve done a great job overall!, Found very few differences to the coding guide, if anything. I think overall a great, nice and neat code!

loyaltypollution committed:
feat: Add template GUI

loyaltypollution committed:
feat: Add GUI for Tom

loyaltypollution committed:
Merge branch 'branch-Level-10'

loyaltypollution committed:
chore: Add assertions to ensure valid inputs in various classes

Currently, several methods across different classes do not have assertions to validate their inputs. This can lead to potential errors and unexpected behavior if invalid inputs are provided.

Adding assertions will help catch invalid inputs early and provide meaningful error messages, improving the robustness of the code.

Add assertions in the following locations:
1. **TaskList.java**:
   - In `addTask` method, assert that the task is not null before adding it to the list.
   - In `removeTask` method, assert that the position is within the valid range before attempting to remove the task.

2. **Deadline.java**:
   - In the constructor, assert that the description and end date are not null.

3. **Event.java**:
   - In the constructor, assert that the description, start date, and end date are not null.

4. **Storage.java**:
   - In `saveFile` method, assert that the tasks list is not null before attempting to save it.

5. **Parser.java**:
   - In `stringToDate` method, assert that the input string is not null and matches the date pattern before attempting to parse it.

6. **Ui.java**:
   - In `showMessage` method, assert that the message is not null.

These assertions are added to ensure that the methods receive valid inputs. This prevents potential runtime errors.

loyaltypollution committed:
chore: Refactor Gui to meet code quality
guidelines

Currently, adding dialog to the MainWindow required MainWindow to
retrieve a response from Tom.java. This leads to Ui needing to store all
messages in an internal buffer before flushing it out in Tom.java.

Adding an interface Ui and an concrete Gui class that interacts with the
MainWindow directly will avoid having multiple levels of abstraction
within a code fragment.

This allows us to meet the code quality standards set out.

This also allows for subsequent CLI implementations of the Ui through
implementation of Ui interface.

loyaltypollution committed:
chore: Remove unneccesary constructors

Currently, there are unneccssary constructors in certain *Command.java
files.

Removing them will maintain the code quality and leave less confusion to
subsequent readers.

This allows us to meet the code quality standards set out.

loyaltypollution committed:
chore: Refactor position validation logic in
TaskList methods

Currently, the position validation logic in the `removeTask` and
`markTask` methods of the `TaskList` class is repeated and not
encapsulated.

This repetition can lead to code duplication and makes it harder to
maintain and update the validation logic.

Refactor the position validation logic into a separate private method
`isValidPosition` and use it in both `removeTask` and `markTask`
methods.

This change reduces code duplication, improves readability, and makes it
easier to maintain and update the validation logic in one place.

Additionally, added Javadocs to the new private method for better
documentation.

loyaltypollution committed:
feat: Refactor TaskList to use Java Streams

The current implementation of TaskList uses traditional loops for
operations like filtering and printing tasks. This can be improved for
readability and conciseness.

Refactor the code to use Java Streams where applicable.

Using Java Streams makes the code more concise and easier to read,
leveraging modern Java features.

No additional changes or dependencies are introduced.

loyaltypollution committed:
Merge pull request #1 from loyaltypollution/branch-A-Assertions

chore: Add assertions to ensure valid inputs in various classes

loyaltypollution committed:
Merge branch 'master' into branch-A-Streams

loyaltypollution committed:
Merge branch 'master' into branch-A-CodeQuality

loyaltypollution committed:
Merge pull request #3 from loyaltypollution/branch-A-Streams

feat: Refactor TaskList to use Java Streams

loyaltypollution committed:
Merge branch 'master' of https://github.com/loyaltypollution/ip into branch-A-CodeQuality

loyaltypollution committed:
feat: Refactor TaskList to use Java Streams

Currently, other branches manage to refactor Tasklist to use Java
Streams. However, toString method has not been refactored.

This commit resolves that discrepency.

loyaltypollution committed:
Merge pull request #2 from loyaltypollution/branch-A-CodeQuality

Adhere to Code Quality standards

loyaltypollution committed:
feat: Add fuzzy matching for task descriptions

Currently, task descriptions are matched exactly with the keyword, which
may not be flexible enough for user searches.

Implement fuzzy matching for task descriptions using Levenshtein
distance to allow for more flexible and user-friendly searches.

Using Levenshtein distance allows for matching keywords with minor typos
or variations, improving the search functionality.

The threshold for the maximum allowed edit distance is set to 2, which
balances flexibility and accuracy.

loyaltypollution committed:
Merge branch 'branch-C-BetterSearch'

loyaltypollution committed:
docs: Add JavaDocs to WordMatch class

loyaltypollution committed:
fix: Fix CheckStyle issues

loyaltypollution committed:
chore: Refactor parsing to subclasses

loyaltypollution committed:
PromptCommand.java: Implement Prompt command

loyaltypollution committed:
feat: Allow responsive GUI prompting

loyaltypollution committed:
feat: Prompt user for more input if needed

loyaltypollution committed:
feat: Use exceptions to pass errors

loyaltypollution committed:
Merge branch 'branch-A-BetterGui'

loyaltypollution committed:
feat: Add Personality features

loyaltypollution committed:
fix: Ensure tom.txt is loaded at start

loyaltypollution committed:
Merge branch 'branch-A-Personality'

loyaltypollution committed:
fix: Fix gui bug in FindCommand

loyaltypollution committed:
docs: Update README with user guide

loyaltypollution committed:
chore: Fix checkstyle issues

loyaltypollution committed:
fix: Create save file if file does not exist

loyaltypollution committed:
chore: SaveCommand throws command exception

loyaltypollution committed:
chore: Seperate command-time and parse exceptions

loyaltypollution committed:
fix: LoadCommand resets list

loyaltypollution committed:
feat: Add WelcomeCommand

loyaltypollution committed:
feat: Add HelpCommand

loyaltypollution committed:
feat: Task events take more parameters

loyaltypollution committed:
fix: Add description for regex failure

loyaltypollution committed:
fix: Refine DeleteCommand response

loyaltypollution committed:
fix: Add commands in list

loyaltypollution committed:
fix: Improve Mark/UnmarkCommand user feedback

loyaltypollution committed:
fix: Improve incomplete pattern user feedback

loyaltypollution committed:
fix: Improve InvalidRegex user feedback

loyaltypollution committed:
docs: Update UserGuide for clarity
